perf(ast, codegen, transformer): avoid allocations when converting `RegExpFlags` to string

overlookmotel opened the pull request:
In various places, we were using `RegExpFlags::to_string` when we want a `&str`. This creates a `String` temporarily, which is an unnecessary allocation.

Add `RegExpFlags::to_inline_string` method which creates an `InlineString` instead, without any allocation. Use this method wherever we convert from `RegExpFlags` to `&str`.

The downside of this change is that `oxc_ast` gains a dependency on `oxc_data_structures`. Personally, I don't think this is so bad, as `oxc_data_structures` is a pretty lightweight crate, containing only quite minimalist data structures. The code is mostly comments! However, what we could do is put each component of `oxc_data_structures` behind it's own feature, so consumers can take only what they need.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9550?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9550** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9550?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9550?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9549** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9549?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9548** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9548?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9547** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9547?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9546** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9546?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9519** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9519?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9522** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9522?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9521** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9521?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9520** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9520?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9516** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9516?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9550).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camchenry commented:
I think the additional dependency is okay personally. It's not too surprising to me that the AST depends on data structures, especially custom ones. We can always turn them into separate crates later.

overlookmotel committed:
perf(ast, codegen, transformer): avoid allocations when converting `RegExpFlags` to string (#9550)

In various places, we were using `RegExpFlags::to_string` when we want a `&str`. This creates a `String` temporarily, which is an unnecessary allocation.

Add `RegExpFlags::to_inline_string` method which creates an `InlineString` instead, without any allocation. Use this method wherever we convert from `RegExpFlags` to `&str`.

The downside of this change is that `oxc_ast` gains a dependency on `oxc_data_structures`. Personally, I don't think this is so bad, as `oxc_data_structures` is a pretty lightweight crate, containing only quite minimalist data structures. The code is mostly comments! However, what we could do is put each component of `oxc_data_structures` behind it's own feature, so consumers can take only what they need.
